Multi Tenant Database

Smightym8 opened the pull request:
- **refactor(db): split ApplicationDbContext into ApplicationDbContext and TenantDbContext for multi-tenancy**
  - Moved user-specific tables into TenantDbContext to support a separate database for each tenant in future multi-tenancy setup.
  - Retained non-user-specific tables in ApplicationDbContext.
  - Updated code to use TenantDbContext for accessing tenant-specific data.
  - Refactored tests to align with the new database structure and TenantDbContext setup.
  - Delete appsettings.Test.json
  

- **feat(tenant): add Tenant class and associate users with tenants**
  - Added `Tenant` class to store tenant information, including connection string for tenant-specific databases.
  - Added `TenantId` field to `User` class to link users to their respective tenants.
  - Implemented tenant association during user registration, storing tenant information without creating a tenant-specific database.
  - Adapted related code.
  

- **feat(tenant): implement creating database for tenant during registration**
  - Added database migration for tenant db
  - Split up database connection string into DatabaseConfiguration
  - Implemented method CreateDatabaseForTenantAsync that creates a new database and a user for the registered tenant. It also created the database schema.
  - Created TenantDbContextFactory to retrieve a TenantDbContext with a dynamically created connection string based on databasename, username and password.
  

- **feat(auth): add username and enable login with username or email**
  - Added `username` field to `User` class to allow login with either `username` or `email`.
  - Refactored login method to check for both `username` and `email`.
  - Refactored tests and added cases to verify login with both `username` and `email`.
  - Added unique constraint for username and email.
  - Refactored tenant database creation to use the new username property as username for the database user.
  

- **feat(tenant): add dynamic connection string creation with error handling for tenant registration**
  - use `NpgsqlConnectionStringBuilder` to dynamically build connection strings for PostgreSQL
  - implement cleanup of created resources (identity user, domain user, tenant) if database creation fails
  

- **feat(tenant): add tenant-aware context handling and update service methods for multi-tenancy**
  - refactor repositories to accept a `Tenant` object to retrieve a `TenantDbContext` dynamically for each tenant.
  - update application services for smart meter management to use tenant information based on the authenticated user from JWT.
  - refactor `UserValidationService` to return the validated user on success.
  

- **refactor(tenant): add creating TenantDbContext with tenant as parameter**
  

- **feat(app): implement dynamic database creation for test user**
  - Removed connection strings from appsettings and implemented dynamic connection string generation
  - Added functionality to create a new database for test users in the development environment, mirroring registration process
  - Implemented seeding of test data into the dynamically created database to ensure consistent test environment
  

- **test: fix application unit tests**
  - Added mocking tenant repository
  

- **test: fix integration tests**
  - Added creating tenant database for test user before each test
  - Added cleanup for tenant database after each test
  - Updated repository tests to pass existing tenant to repository
  

- **refactor(tenant): change exceptions and split up CreateDatabaseForTenantAsync method**
  - Introduced TenantNotFoundException and use it instead of Exception.
  - Split method `CreateDatabaseForTenantAsync` into smaller methods.
  

- **feat(middleware): centralize userId extraction in JwtClaimMiddleware**
  - Refactor application services to retrieve userId from HttpContext via IHttpContextAccessor
  - Remove userId argument from service methods
  - Update all relevant endpoints and services to rely on HttpContextAccessor for userId retrieval
  - Implement JwtClaimMiddleware to parse JWT token and store userId in HttpContext.Items
  - Update unit tests to reflect changes in userId handling
  

- **feat(tenant): implement TenantContextService to manage tenant resolution**
  - Introduced TenantContextService with GetCurrentTenantAsync to retrieve current tenant based on userId from JWT
  - Refactored application services to call TenantContextService for tenant retrieval, removing direct userId and database calls
  - Removed UserId field from SmartMeter model, as it is now implied by tenant-user relationship
  - Recreated migrations for TenantDbContext to reflect updated schema
  - Updated unit and integration tests
  

- **refactor: simplify TenantDbContext management and resolve merge issues**
  - Fixed code conflicts from merging main branch into feature branch
  - Refactored repository methods to remove the need to pass tenant information explicitly
  - Registered TenantDbContext per request
  - Streamlined repository methods by removing inline TenantDbContext instantiation
  

- **style: format code with dotnet format**

Smightym8 committed:
refactor(db): split ApplicationDbContext into ApplicationDbContext and TenantDbContext for multi-tenancy

- Moved user-specific tables into TenantDbContext to support a separate database for each tenant in future multi-tenancy setup.
- Retained non-user-specific tables in ApplicationDbContext.
- Updated code to use TenantDbContext for accessing tenant-specific data.
- Refactored tests to align with the new database structure and TenantDbContext setup.
- Delete appsettings.Test.json

Smightym8 committed:
feat(tenant): add Tenant class and associate users with tenants

- Added `Tenant` class to store tenant information, including connection string for tenant-specific databases.
- Added `TenantId` field to `User` class to link users to their respective tenants.
- Implemented tenant association during user registration, storing tenant information without creating a tenant-specific database.
- Adapted related code.

Smightym8 committed:
feat(tenant): implement creating database for tenant during registration

- Added database migration for tenant db
- Split up database connection string into DatabaseConfiguration
- Implemented method CreateDatabaseForTenantAsync that creates a new database and a user for the registered tenant. It also created the database schema.
- Created TenantDbContextFactory to retrieve a TenantDbContext with a dynamically created connection string based on databasename, username and password.

Smightym8 committed:
feat(auth): add username and enable login with username or email

- Added `username` field to `User` class to allow login with either `username` or `email`.
- Refactored login method to check for both `username` and `email`.
- Refactored tests and added cases to verify login with both `username` and `email`.
- Added unique constraint for username and email.
- Refactored tenant database creation to use the new username property as username for the database user.

Smightym8 committed:
feat(tenant): add dynamic connection string creation with error handling for tenant registration

- use `NpgsqlConnectionStringBuilder` to dynamically build connection strings for PostgreSQL
- implement cleanup of created resources (identity user, domain user, tenant) if database creation fails

Smightym8 committed:
feat(tenant): add tenant-aware context handling and update service methods for multi-tenancy

- refactor repositories to accept a `Tenant` object to retrieve a `TenantDbContext` dynamically for each tenant.
- update application services for smart meter management to use tenant information based on the authenticated user from JWT.
- refactor `UserValidationService` to return the validated user on success.

Smightym8 committed:
refactor(tenant): add creating TenantDbContext with tenant as parameter

Smightym8 committed:
feat(app): implement dynamic database creation for test user

- Removed connection strings from appsettings and implemented dynamic connection string generation
- Added functionality to create a new database for test users in the development environment, mirroring registration process
- Implemented seeding of test data into the dynamically created database to ensure consistent test environment

Smightym8 committed:
test: fix application unit tests

- Added mocking tenant repository

Smightym8 committed:
test: fix integration tests

- Added creating tenant database for test user before each test
- Added cleanup for tenant database after each test
- Updated repository tests to pass existing tenant to repository

Smightym8 committed:
refactor(tenant): change exceptions and split up CreateDatabaseForTenantAsync method

- Introduced TenantNotFoundException and use it instead of Exception.
- Split method `CreateDatabaseForTenantAsync` into smaller methods.

Smightym8 committed:
feat(middleware): centralize userId extraction in JwtClaimMiddleware

- Refactor application services to retrieve userId from HttpContext via IHttpContextAccessor
- Remove userId argument from service methods
- Update all relevant endpoints and services to rely on HttpContextAccessor for userId retrieval
- Implement JwtClaimMiddleware to parse JWT token and store userId in HttpContext.Items
- Update unit tests to reflect changes in userId handling

Smightym8 committed:
feat(tenant): implement TenantContextService to manage tenant resolution

- Introduced TenantContextService with GetCurrentTenantAsync to retrieve current tenant based on userId from JWT
- Refactored application services to call TenantContextService for tenant retrieval, removing direct userId and database calls
- Removed UserId field from SmartMeter model, as it is now implied by tenant-user relationship
- Recreated migrations for TenantDbContext to reflect updated schema
- Updated unit and integration tests

Smightym8 committed:
Merge remote-tracking branch 'origin/main' into feature/setup-multi-tenant-database

# Conflicts:
#	src/SMAIAXBackend.API/ApplicationConfigurations/ServiceExtensions.cs
#	src/SMAIAXBackend.API/Endpoints/SmartMeter/GetSmartMetersEndpoint.cs
#	src/SMAIAXBackend.Application/Services/Implementations/AuthenticationService.cs
#	src/SMAIAXBackend.Application/Services/Implementations/SmartMeterCreateService.cs
#	src/SMAIAXBackend.Application/Services/Interfaces/ISmartMeterCreateService.cs
#	src/SMAIAXBackend.Domain/Repositories/ISmartMeterRepository.cs
#	src/SMAIAXBackend.Infrastructure/Repositories/SmartMeterRepository.cs
#	tests/SMAIAXBackend.Application.UnitTests/SmartMeterCreateServiceTests.cs
#	tests/SMAIAXBackend.IntegrationTests/TestBase.cs

Smightym8 committed:
refactor: simplify TenantDbContext management and resolve merge issues

- Fixed code conflicts from merging main branch into feature branch
- Refactored repository methods to remove the need to pass tenant information explicitly
- Registered TenantDbContext per request
- Streamlined repository methods by removing inline TenantDbContext instantiation

Smightym8 committed:
style: format code with dotnet format

Smightym8 committed:
build: add Microsoft.AspNetCore.Http.Abstractions via NuGet

- Replaced file system reference to Microsoft.AspNetCore.Http.Abstractions with a NuGet package
- Ensures consistent package management and simplifies dependency resolution

Smightym8 committed:
fix(app): resolve SonarCloud-reported issues

Essiga commented on src/SMAIAXBackend.API/ApplicationConfigurations/AuthenticationServiceExtensions.cs:
```diff
@@ -17,6 +17,7 @@ public static void AddIdentityConfigurations(this IServiceCollection services)
     {
         services.AddIdentity<IdentityUser, IdentityRole>(options =>
             {
+                options.User.AllowedUserNameCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
```
we should also allow numbers 0-9

Smightym8 committed:
feat(auth): add digits 0-9 to AllowedUserNameCharacters

- Updated `AllowedUserNameCharacters` to include numeric characters (0-9)
- Expands the valid character set for usernames to allow digits

Essiga commented on src/SMAIAXBackend.API/Program.cs:
```diff
@@ -46,20 +51,42 @@
+        await deleteUserCommand.ExecuteNonQueryAsync();
+        await applicationDbContext.Database.CloseConnectionAsync();
+    }
+    await tenantRepository.CreateDatabaseForTenantAsync("tenant_1_db", "johndoe", "P@ssw0rd");
```
shouldn't we read username and password from a file as opposed to having it hardcoded? Even if it's just for the test db?

Essiga reviewed (approved):
lgtm

Smightym8 committed:
refactor(app): inject IConfiguration into ApplicationDbContext for configurable test data seeding

- Added `IConfiguration` injection to `ApplicationDbContext` to access appsettings values
- Modified `SeedTestData` to retrieve test user data (username, password, database) from configuration instead of hard-coded values
- Modified Program.cs to retrieve test user data (username, password, database) from configuration instead of hard-coded values
- Improved flexibility for development environments by making test user data configurable

Smightym8 committed:
Merge remote-tracking branch 'origin/main' into feature/setup-multi-tenant-database

# Conflicts:
#	src/SMAIAXBackend.API/ApplicationConfigurations/RepositoryExtensions.cs
#	src/SMAIAXBackend.API/ApplicationConfigurations/ServiceExtensions.cs
#	src/SMAIAXBackend.Application/Services/Implementations/SmartMeterListService.cs
#	src/SMAIAXBackend.Domain/Model/Entities/SmartMeter.cs
#	src/SMAIAXBackend.Infrastructure/DbContexts/ApplicationDbContext.cs
#	src/SMAIAXBackend.Infrastructure/Repositories/SmartMeterRepository.cs
#	tests/SMAIAXBackend.Application.UnitTests/SmartMeterListServiceTests.cs
#	tests/SMAIAXBackend.Domain.UnitTests/SmartMeterTests.cs
#	tests/SMAIAXBackend.IntegrationTests/IntegrationTestSetup.cs

Smightym8 committed:
fix(app): resolve merge conflicts and update policy code for multi-tenancy support

- Refactored policy and policy request code to support multi-tenancy after merge
- Regenerated migrations for `TenantDbContext` to align with latest schema changes
- Fixed issues introduced by merging main branch into the current branch
